addons/l10n_sa_edi/models/res_company.py

```diff
@@ -0,0 +1,111 @@
+        """
+        private_key = ec.generate_private_key(ec.SECP256K1, default_backend())
+        return private_key.private_bytes(
+            encoding=serialization.Encoding.PEM,
+            format=serialization.PrivateFormat.TraditionalOpenSSL,
+            encryption_algorithm=serialization.NoEncryption()
+        )
+
+    l10n_sa_private_key = fields.Binary("ZATCA Private key", attachment=False, groups="base.group_erp_manager",
+                                        copy=False,
+                                        help="The private key used to generate the CSR and obtain certificates",
+                                        default=_l10n_sa_compute_private_key)
+
+    l10n_sa_api_mode = fields.Selection(
+        [('sandbox', 'Sandbox'), ('preprod', 'Simulation (Pre-Production)'), ('prod', 'Production')],
+        help="Specifies which API the system should use", required=True,
+        default='sandbox', copy=False)
+
+    l10n_sa_edi_building_number = fields.Char(compute='_compute_address',
+                                              inverse='_l10n_sa_edi_inverse_building_number')
+    l10n_sa_edi_plot_identification = fields.Char(compute='_compute_address',
+                                                  inverse='_l10n_sa_edi_inverse_plot_identification')
+    l10n_sa_edi_neighborhood = fields.Char(compute='_compute_address', inverse='_l10n_sa_edi_inverse_neighborhood')
+
+    l10n_sa_additional_identification_scheme = fields.Selection(
+        related='partner_id.l10n_sa_additional_identification_scheme', readonly=False)
+    l10n_sa_additional_identification_number = fields.Char(
+        related='partner_id.l10n_sa_additional_identification_number', readonly=False)
+
+    def write(self, vals):
+        for company in self:
+            if company.l10n_sa_api_mode == 'prod' and 'l10n_sa_api_mode' in vals and vals['l10n_sa_api_mode'] != 'prod':
+                raise UserError("You cannot change the ZATCA Submission Mode once it has been set to Production")
+        return super().write(vals)
+
+    def _get_company_address_field_names(self):
+        """ Override to add ZATCA specific address fields """
+        return super()._get_company_address_field_names() + \
+            ['l10n_sa_edi_building_number', 'l10n_sa_edi_plot_identification', 'l10n_sa_edi_neighborhood']
+
+    def _l10n_sa_edi_inverse_building_number(self):
+        for company in self:
+            company.partner_id.l10n_sa_edi_building_number = company.l10n_sa_edi_building_number
+
+    def _l10n_sa_edi_inverse_plot_identification(self):
+        for company in self:
+            company.partner_id.l10n_sa_edi_plot_identification = company.l10n_sa_edi_plot_identification
+
+    def _l10n_sa_edi_inverse_neighborhood(self):
+        for company in self:
+            company.partner_id.l10n_sa_edi_neighborhood = company.l10n_sa_edi_neighborhood
+
+    def _l10n_sa_get_csr_invoice_type(self):
```

Comment:
Should remain on the company, because this is used to determine whether we'll submit standard and/or simplified invoices, and should not change from journal to journal. ZATCA might also expand on the invoice_type later on to add more types